### PR TITLE
Floating numbers

### PR DESCRIPTION
**Pull request by AJenbo**
This replaces #4580 as an initial implementation for floating numbers. The cleanups that have happened since it was originally created has helped simplify a few things. Additionally it replaces `enum missile_resistance` with `enum class DamageType` an breaks up the logic for calculating the worldtile to screen conversion which allows it to now have simpler and more accurate calculation for where to display the numbers. The code is still based on @qndel original work and retains him as the author.

Font colors are also reuses the existing system rather then implementing a new scheme. D1 Graphics Tool was helpful here for editing and previewing the TRN files.

This version does not have the many options and is instead simply on/off. I think it would be ok to introduce a few additional options like only showing number for high damage values. Maybe also toggles for healing and XP. But that can be done as a follow up once we have the basics merged. For that I would also prefer if we could spend some time breaking up the settings in to multiple pages and cleaning out some of other options.

One remaning issue is that the number will continue to move even when the game is paused since they are based on real time and not game time.

### CODE REVIEW

**Comment by Trihedraf** [comment]
Just stating my preference, but the I don't like the numbers going out in different directions. The other PR with all the options there was the angle options which could be set to make the numbers go straight up which is much cleaner looking to me.

**Comment by julealgon** [comment]
> Additionally it replaces `enum missile_resistance` with `enum class DamageType` an breaks up the logic for calculating the worldtile to screen conversion which allows it to now have simpler and more accurate calculation for where to display the numbers.

Would be nice if you could split these things into at least a separate commit.

**Comment by AJenbo** [comment]
With the prep PRs now merged this is ready for review

File: Source/monster.cpp [hunk 1]
```diff
@@ -3568,8 +3569,10 @@ void AddDoppelganger(Monster &monster)
 	}
 }
 
-void ApplyMonsterDamage(Monster &monster, int damage)
+void ApplyMonsterDamage(DamageType dmgType, Monster &monster, int damage)
```
Comment (StephenCWills) [inline Source/monster.cpp:3572]: Any reason we shouldn't spell all these out (`damageType`)?

**Review by StephenCWills — CHANGES_REQUESTED** [review]

File: Source/qol/floatingnumbers.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+		{ -100, -100 },
+	};
+
+	Displacement endOffset = goodAngles[rand() % 3];
+
+	if (damageToPlayer)
+		endOffset = -endOffset;
+
+	for (auto &num : FloatingQueue) {
+		if (num.reverseDirection == damageToPlayer && num.type == type && num.index == index && (SDL_GetTicks() - (int)num.lastMerge) <= 100) {
```
Comment (glebm) [inline Source/qol/floatingnumbers.cpp:106]: Use `static_cast`

File: Source/qol/floatingnumbers.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+		DrawString(out, floatingNum.text, Rectangle { screenPosition, { lineWidth, 0 } }, floatingNum.style);
+	}
+
+	ClearExpiredNumbers();
+}
+
+void ClearFloatingNumbers()
+{
+	srand(time(NULL));
```
Comment (glebm) [inline Source/qol/floatingnumbers.cpp:193]: Use `nullptr`

File: Source/engine/render/scrollrt.cpp [hunk 4]
```diff
@@ -984,101 +985,94 @@ Displacement tileShift;
...
+	if (myPlayer.isWalking())
+		offset += GetOffsetForWalking(myPlayer.AnimInfo, myPlayer._pdir, true);
+
+	position += tileShift;
+
+	// Skip rendering parts covered by the panels
+	if (CanPanelsCoverView() && (IsLeftPanelOpen() || IsRightPanelOpen())) {
+		int multiplyer = (*sgOptions.Graphics.zoom) ? 1 : 2;
+		position += Displacement(Direction::East) * multiplyer;
+		offset.deltaX += -TILE_WIDTH * multiplyer / 2 / 2;
```
[the remark this answers is not in the document]
Comment (AJenbo) [reply Source/engine/render/scrollrt.cpp:1002]: one for getting half the tile with, and one for half the resolution. I'm pretty sure the compiler will optimize it.

File: Source/qol/floatingnumbers.cpp (hunk 2 above)
Comment (StephenCWills) [reply Source/qol/floatingnumbers.cpp:106]:
```suggestion
		if (num.reverseDirection == damageToPlayer && num.type == type && num.index == index && (SDL_GetTicks() - static_cast<int>(num.lastMerge)) <= 100) {
```

File: Source/qol/floatingnumbers.cpp (hunk 3 above)
Comment (StephenCWills) [reply Source/qol/floatingnumbers.cpp:193]:
```suggestion
	srand(time(nullptr));
```

**Review by StephenCWills — DISMISSED** [review]
Somehow, I broke all the GitHub Actions jobs with an error that makes no sense to me, but otherwise the PR looks good now. The only suggestion I have left is to maybe change `enableFloatingNumbers` to `floatingNumbersMode` with an enum for `None`, `Up`, and `Random`. That could satisfy @Trihedraf's request without introducing another separate setting.

**Comment by AJenbo** [comment]
Probably because it spans the border between us upgrading the CI version

**Comment by AJenbo** [comment]
> The only suggestion I have left is to maybe change `enableFloatingNumbers` to `floatingNumbersMode` with an enum for `None`, `Up`, and `Random`. That could satisfy @Trihedraf's request without introducing another separate setting.

I'm not dismissing this, but going to wait with it for a follow up

File: Source/monster.cpp (hunk 1 above)
Comment (AJenbo) [reply Source/monster.cpp]: Fixed

File: Source/missiles.cpp [hunk 5]
```diff
@@ -245,15 +246,15 @@ bool MonsterMHit(int pnum, int monsterId, int mindam, int maxdam, int dist, Miss
 		dam >>= 2;
 
 	if (&player == MyPlayer)
-		ApplyMonsterDamage(monster, dam);
+		ApplyMonsterDamage(MissilesData[static_cast<int8_t>(t)].damageType, monster, dam);
```
Comment (StephenCWills) [inline Source/missiles.cpp:249]:
```suggestion
		ApplyMonsterDamage(missileData.damageType, monster, dam);
```

**Review by StephenCWills — CHANGES_REQUESTED** [review]

**Review by StephenCWills — APPROVED** [review]

**Pull request merged**